### PR TITLE
Work order receipt APIs implementation for direct model

### PR DESCRIPTION
**Pull request by Ram-srini**
1. Work order receipt API's create, update, retrieve and update retrieve
are implemented. Updating receipt is done by enclave manager based on the
work order execution status.
2. Modify work receipt jrpc wrapper as per the new EEA spec changes
3. Update to echo client with receipt APIs to test end to end flow.

Signed-off-by: Ramakrishna Srinivasamurthy <ramakrishna.srinivasamurthy@intel.com>

### CODE REVIEW

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py [hunk 1]
```diff
@@ -62,6 +65,11 @@ def __init__(self, config, signup_data, measurements):
         self.proof_data_type = config.get("WorkerConfig")["ProofDataType"]
         self.proof_data = signup_data.proof_data
 
+        # Key pair for work order receipt signing
+        # This is temporary aaproach
```
Comment (danintel) [inline examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: s/aaproach/approach/

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py [hunk 2]
```diff
@@ -203,18 +215,51 @@ def process_work_orders(self, kv_helper):
...
 
         # end of for loop
 
+    # -----------------------------------------------------------------
+
+    def __update_receipt(self, kv_helper, wo_id, wo_json_resp):
+        """
+        Update the existing work order receipt with the status as in wo_json_
+        Parameters:
+            - kv_helper is lmdb instatance to access database
```
Comment (danintel) [inline examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: s/instatance/instance/

File: examples/common/python/work_order_receipt/work_order_receipt_request.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import random
+
+import crypto.crypto as crypto
+import utility.signature as signature
+from error_code.error_status import ReceiptCreateStatus as ReceiptCreateStatus
+
+class WorkOrderReceiptRequest():
+    """
+    Class to create work order receipt APIs like
+    create, update, retrieve and loookup
```
Comment (danintel) [inline examples/common/python/work_order_receipt/work_order_receipt_request.py]: s/loookup/lookup/

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py [hunk 4]
```diff
@@ -38,6 +43,12 @@ def __init__(self, kv_helper):
         self.kv_helper = kv_helper
         self.receipt_pool = []
         self.__workorder_receipt_on_boot()
+        # Special index value to fetch last update to receipt
```
Comment (danintel) [inline examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Add `0xFFFFFFFF` to the comment.

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py [hunk 5]
```diff
@@ -48,91 +59,222 @@ def __workorder_receipt_on_boot(self):
...
-                    "Receipt created successfully")
-            else:
-                response = utility.create_error_response(
-                    WorkOrderStatus.INVALID_PARAMETER_FORMAT_OR_VALUE, jrpc_id,
-                    "Work order does not exists. Hence invalid parameter")
+        wo_request = self.kv_helper.get("wo-requests", wo_id)
+        if wo_request is None:
+            raise JSONRPCDispatchException(
+                JRPCErrorCodes.INVALID_PARAMETER_FORMAT_OR_VALUE,
+                "Work order id does not exists, "
```
Comment (danintel) [inline examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: s/exists/exist/

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py [hunk 6]
```diff
@@ -229,7 +274,7 @@ def create_enclave_signup_data():
 
     return enclave_signup_data
 
-# -----------------------------------------------------------------
+    # -----------------------------------------------------------------
```
Comment (danintel) [inline examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: This should not be indented. Was this a mistake? Or on purpose?

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py [hunk 7]
```diff
@@ -370,15 +406,16 @@ def start_enclave_manager(config):
...
-            enclave_manager.process_work_orders(kv_helper)
-            logger.info("Enclave manager sleeping for %d secs", sleep_interval)
-            time.sleep(sleep_interval)
-    except:
-        logger.error("Error while processing work-order. Shutting down enclave manager")
-        exit(1)
+        enclave_manager.process_work_orders(kv_helper)
+        logger.info("Enclave manager sleeping for %d secs", sleep_interval)
+        time.sleep(sleep_interval)
+#    except Exception as inst:
```
Comment (danintel) [inline examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: Remove this commented out code, or explain why it should be left here.

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py (hunk 4 above)
Comment (Ram-srini) [reply examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Done

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py (hunk 5 above)
Comment (Ram-srini) [reply examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Done

File: examples/common/python/work_order_receipt/work_order_receipt_request.py (hunk 3 above)
Comment (Ram-srini) [reply examples/common/python/work_order_receipt/work_order_receipt_request.py]: Done

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py (hunk 1 above)
Comment (Ram-srini) [reply examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: Done

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py (hunk 2 above)
Comment (Ram-srini) [reply examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: Done

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py (hunk 6 above)
Comment (Ram-srini) [reply examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: Corrected

**Comment by pankajgoyal2** [comment]
in examples/apps/echo/client/echo_client.py, line 46, TCFHOME, the default is set to 2 directories up (../../) i.e. examples/common folder. Doesn't it has to be 4 levels up (../../../../)?

File: examples/common/python/utility/signature.py [hunk 8]
```diff
@@ -315,4 +315,106 @@ def verify_signature(self, input_json, verification_key):
...
+            return SignatureStatus.INVALID_SIGNATURE_FORMAT
+
+#---------------------------------------------------------------------------------------------
+    def calculate_request_hash(self, input_json):
+        """
+        Function to create the work order reuest hash
+        as defined in EEA spec 6.1.8.1
+        Parameters:
+            - input_json is dictionary contains payload returned by the
+              WorkOrderReceiptRetrieve API as define EEA spec 6.1.1
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/common/python/utility/signature.py]: Comment was wrong input_json is payload work order request payload as defined in 6.1.1. I have changed the comments.

File: examples/common/python/work_order_receipt/work_order_receipt_request.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        wo_resp_hash = crypto.compute_message_hash(wo_resp_bytes)
+        wo_resp_hash_str = crypto.byte_array_to_hex(wo_resp_hash)
+        receipt_status = None
+
+        public_key = signing_key.GetPublicKey().Serialize()
+        wo_id = None
+        if "error" in wo_resp and \
+            wo_resp["error"]["code"] != WorkOrderStatus.PENDING.value:
+            receipt_status = ReceiptCreateStatus.FAILED.value
+            wo_id = wo_resp["error"]["data"]
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/common/python/work_order_receipt/work_order_receipt_request.py]: If it is an error in work order request then work order status payload should have data which contains workOrderId. It is defined in section 6.1.3

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py [hunk 10]
```diff
@@ -35,311 +39,399 @@ def __init__(self, kv_helper):
...
+                        return False, "invalid data parameter for " + key
+                elif key in ["workOrderRequestHash", "requesterSignature"]:
+                    try:
+                        base64.b64decode(wo_receipt_req["params"][key])
+                    except Exception as e:
+                        return False, \
+                            "Invalid data format for " + key
+        receipt_type = wo_receipt_req["params"]["receiptCreateStatus"]
+        try:
+            receipt_enum_type = ReceiptCreateStatus(receipt_type)
```
Comment (manju956) [inline examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Can you remove receipt_enum_type variable since its not being used?

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py [hunk 11]
```diff
@@ -35,311 +39,399 @@ def __init__(self, kv_helper):
...
+        # Verify work order request is calculated properly or not.
+        wo_req_hash = signature_obj.calculate_request_hash(json.loads(wo_request))
+        if wo_req_hash != wo_receipt_req["params"]["workOrderRequestHash"]:
+            return False, "Work order request hash does not match"
+        # Verify requester signature with signing key in the request
+        status = signature_obj.verify_create_receipt_signature(wo_receipt_req)
+        if status != SignatureStatus.PASSED:
+            return False, "Receipt create requester signature does not match"
+        # If all parameters are verified in the request
+        return True, ""
```
Comment (manju956) [inline examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Can we add message(say success) after True value?

File: examples/common/python/work_order_receipt/work_order_receipt_request.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        else:
+            receipt_status = ReceiptCreateStatus.PROCESSED.value
+            wo_id = wo_resp["result"]["workOrderId"]
+        
+        wo_receipt_update = {
+            "workOrderId": wo_id,
+            "updaterId": wo_id,
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/common/python/work_order_receipt/work_order_receipt_request.py]: Change it to worker_id for now.

File: examples/common/python/utility/signature.py [hunk 13]
```diff
@@ -315,4 +315,106 @@ def verify_signature(self, input_json, verification_key):
...
+            input_json_params["workerServiceId"] + \
+            input_json_params["workerId"] + \
+            input_json_params["requesterId"] + \
+            str(input_json_params["receiptCreateStatus"]) + \
+            input_json_params["workOrderRequestHash"] + \
+            input_json_params["requesterGeneratedNonce"]
+        concat_hash = bytes(concat_string, "UTF-8")
+        final_hash = crypto.compute_message_hash(concat_hash)
+        signature = input_json_params["requesterSignature"]
+        verification_key = input_json_params["receiptVerificationKey"]
```
Comment (manju956) [inline examples/common/python/utility/signature.py]: both verify_receipt_create_signature and verify_receipt_update_signature look similar except calculating hash on different params. Can we create a utility function for this?

File: examples/common/python/work_order_receipt/__init__.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2018 Intel Corporation
```
Comment (manju956) [inline examples/common/python/work_order_receipt/__init__.py]: Update copyright year to 2019

File: examples/common/python/work_order_receipt/work_order_receipt_request.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+        Returns jrpc request of type dictionary
+        """
+        final_hash_str = self.sig_obj.calculate_request_hash(wo_request)
+        wo_request_params = wo_request["params"]
+        worker_id = wo_request_params["workerId"]
+        requester_nonce = nonce
+        if nonce is None:
+            requester_nonce = str(random.randint(1, 10**10))
+        public_key =  signing_key.GetPublicKey().Serialize()
+        wo_receipt_reqest = {
```
Comment (manju956) [inline examples/common/python/work_order_receipt/work_order_receipt_request.py]: typo in variable name. change wo_receipt_reqest -> wo_receipt_request here and wherever used.

File: examples/common/python/work_order_receipt/work_order_receipt_request.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+            receipt_status = ReceiptCreateStatus.FAILED.value
+            wo_id = wo_resp["error"]["data"]
+        else:
+            receipt_status = ReceiptCreateStatus.PROCESSED.value
+            wo_id = wo_resp["result"]["workOrderId"]
+        
+        wo_receipt_update = {
+            "workOrderId": wo_id,
+            "updaterId": wo_id,
+            "updateType": receipt_status,
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/common/python/work_order_receipt/work_order_receipt_request.py]: Spec is defined with 0 to 5 defined values and 5 to 255 are reserved and 255 for any states and above 255 are application specific. Since  our reference implementation EM is updating it knows about 0 to 4.

File: examples/common/python/work_order_receipt/work_order_receipt_request.py [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+            "updateData": wo_resp_hash_str,
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/common/python/work_order_receipt/work_order_receipt_request.py]: Same as above.

File: examples/common/python/work_order_receipt/work_order_receipt_request.py [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+            wo_receipt_reqest["workOrderRequestHash"] + \
+            wo_receipt_reqest["requesterGeneratedNonce"]
+        wo_receipt_bytes = bytes(wo_receipt_str, "UTF-8")
+        wo_receipt_hash = crypto.compute_message_hash(wo_receipt_bytes)
+        status, wo_receipt_sign = self.sig_obj.generate_signature(
+            wo_receipt_hash,
+            signing_key
+        )
+        if status == False:
+            wo_receipt_reqest["requesterSignature"] = ""
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/common/python/work_order_receipt/work_order_receipt_request.py]: Done

File: examples/apps/generic_client/generic_client.py [hunk 19]
```diff
@@ -283,9 +320,8 @@ def Main(args=None):
 	logger.info("Work order get result : {}\n ".format(
 		json.dumps(res, indent=4)
 	))
-	
+	sig_obj = signature.ClientSignature()
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/apps/generic_client/generic_client.py]: it is used outside of if condition.

File: examples/common/python/connectors/direct/tcs_listener/tcs_listener.py [hunk 20]
```diff
@@ -102,29 +109,14 @@ def _process_request(self, input_json_str):
         try:
             input_json = json.loads(input_json_str)
         except:
-            response['error']['message'] = 'Error: Improper Json. Unable to load'
+            response = {
+                "error": {
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/common/python/connectors/direct/tcs_listener/tcs_listener.py]: If request itself is not a valid json then, how do we get the jrpc id from the request.

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py (hunk 11 above)
Comment (Ram-srini) [reply examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: we are not using the second argument if it success. In case error second argument will have different error messages.

File: examples/common/python/utility/signature.py (hunk 13 above)
Comment (Ram-srini) [reply examples/common/python/utility/signature.py]: No both are different.

File: examples/common/python/work_order_receipt/work_order_receipt_request.py (hunk 15 above)
Comment (Ram-srini) [reply examples/common/python/work_order_receipt/work_order_receipt_request.py]: Corrected.

File: examples/common/python/work_order_receipt/__init__.py (hunk 14 above)
Comment (Ram-srini) [reply examples/common/python/work_order_receipt/__init__.py]: Corrected.

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py (hunk 10 above)
Comment (Ram-srini) [reply examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Added it's usage.

File: examples/apps/generic_client/generic_client.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/apps/generic_client/generic_client.py]: Usage is outside of if condition.

File: examples/apps/echo/client/echo_client.py [hunk 21]
```diff
@@ -251,6 +252,40 @@ def Main(args=None):
...
+		wo_request,
+		ReceiptCreateStatus.PENDING.value,
+		private_key
+	)
+	logger.info("Work order create receipt request : {} \n \n ".format(
+		json.dumps(wo_create_receipt, indent=4)
+	))
+	# Submit work order create receipt jrpc request
+	wo_receipt_resp = wo_receipt_instance.work_order_receipt_create(
+		wo_create_receipt["workOrderId"],
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply examples/apps/echo/client/echo_client.py]: No. It is generalized for both direct and proxy using base class. Few arguments are not there in proxy model, those are can be optional arguments in proxy.

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py [hunk 22]
```diff
@@ -35,311 +39,399 @@ def __init__(self, kv_helper):
...
-            if len(updater_objects) > 0:
-                # Updater Object is sorted based on index and then last index is chosen
-                index = int(sorted(updater_objects.keys())[-1]) + 1
+            # Receipt is created, validate update receipt request
+            status, err_msg = self.__validate_work_order_receipt_update_req(
+                    input_value
+                )
+            if status == True:
+                # Load previous updates to receipt
+                updated_receipt = None
```
Comment (rranjan3) [inline examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: This initialization  might be redundant.

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py [hunk 23]
```diff
@@ -35,311 +39,399 @@ def __init__(self, kv_helper):
...
+    def __validate_work_order_receipt_create_req(self, wo_receipt_req, wo_request):
+        """
+        Function to validate the work order receipt create request parameters
+        Parameters:
+            - wo_receipt_req is work order receipt request as dictionary
+            - wo_request is string containing jrpc work order request
+        Returns - tuple containing validation status(Boolean) and error message(string)
+        """
+        # Valid parameters list
+        valid_params = ["workOrderId", "workerServiceId", "workerId", "requesterId",
```
Comment (rranjan3) [inline examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: We should consider putting all such strings to a constants file or declare at the top of a py file otherwise. Could be taken up as a separate PR for the complete codebase.

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py [hunk 24]
```diff
@@ -35,311 +39,399 @@ def __init__(self, kv_helper):
...
+
 # ------------------------------------------------------------------------------------------------
 
     def __workorder_receipt_on_boot(self):
         """
         Function to perform on-boot process of work order handler
         """
-
-        self.receipt_pool = self.kv_helper.lookup("wo-receipts")
+        # TODO: Need to be implemented the on boot flow.
```
Comment (rranjan3) [inline examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Minor typo.

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py [hunk 25]
```diff
@@ -35,311 +39,399 @@ def __init__(self, kv_helper):
...
+                    )
+                else:
+                    raise JSONRPCDispatchException(
+                        JRPCErrorCodes.INVALID_PARAMETER_FORMAT_OR_VALUE,
+                        err_msg
+                    )
+            else:
+                raise JSONRPCDispatchException(
+                    JRPCErrorCodes.INVALID_PARAMETER_FORMAT_OR_VALUE,
+                    "Work order receipt is already exists. Hence invalid parameter"
```
Comment (rranjan3) [inline examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Typo: Consider removing extra 'is'

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py [hunk 26]
```diff
@@ -35,311 +39,399 @@ def __init__(self, kv_helper):
...
+                if last_update["updateType"] == ReceiptCreateStatus.PROCESSED.value:
+                    if input_value["params"]["updateType"] in [
+                        ReceiptCreateStatus.PENDING.value,
+                        ReceiptCreateStatus.FAILED.value,
+                        ReceiptCreateStatus.REJECTED.value
+                    ]:
+                        raise JSONRPCDispatchException(
+                            JRPCErrorCodes.INVALID_PARAMETER_FORMAT_OR_VALUE,
+                            "Current receipt status is set to processed, "
+                            "setting it to status pending, failed or rejected are not allowed"
```
Comment (rranjan3) [inline examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Consider replacing with -
"setting it to status "+input_value["params"]["updateType"]+" are not allowed"

File: examples/apps/echo/client/echo_client.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (manojgop) [reply examples/apps/echo/client/echo_client.py]: Can we address these changes in separate PR ?

File: examples/apps/echo/client/echo_client.py [hunk 27]
```diff
@@ -29,274 +29,339 @@
...
-	worker_id = options.worker_id
-	message = options.message
-	if options.message is None or options.message == "":
-		message = "Test Message"
-
-	# Initializing Worker Object
-	worker_obj = worker.SGXWorkerDetails()
+TCFHOME = os.environ.get("TCF_HOME", "../../../../")
+
+def ParseCommandLine(args):
```
Comment (manojgop) [inline examples/apps/echo/client/echo_client.py]: This file has too many changes which is unrelated to Work order receipt implementation. Can you please move the formatting related changes to separate PR

File: examples/apps/generic_client/generic_client.py [hunk 28]
```diff
@@ -41,284 +42,334 @@
...
-		config["tcf"]["json_rpc_uri"] = uri
-
-	address = options.address
-	if address:
-		if mode == "listing":
-			config["ethereum"]["direct_registry_contract_address"] = \
-				address
-		elif mode == "registry":
-			logger.error("\n Only Worker registry listing address is supported." +
+    global worker_obj
```
Comment (manojgop) [inline examples/apps/generic_client/generic_client.py]: This file has too many changes which is unrelated to Work order receipt implementation. Can you please move the formatting related changes to separate PR.

File: examples/apps/echo/client/echo_client.py (hunk 27 above)
Comment (Ram-srini) [reply examples/apps/echo/client/echo_client.py]: Formatting changes are related to remove tab and space mix usages. I don't want to remove it to separate PR.

File: examples/apps/generic_client/generic_client.py (hunk 28 above)
Comment (Ram-srini) [reply examples/apps/generic_client/generic_client.py]: Formatting changes are related to remove tab and space mix usages. I don't want to move it to separate PR.

File: examples/common/python/connectors/direct/tcs_listener/tcs_listener.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (manju956) [reply examples/common/python/connectors/direct/tcs_listener/tcs_listener.py]: Ok

File: examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py (hunk 11 above)
Comment (manju956) [reply examples/common/python/connectors/direct/tcs_listener/tcs_workorder_receipt_handler.py]: Ok
